Mathlib/LinearAlgebra/Matrix/SchurTriangulation.lean

```diff
@@ -0,0 +1,300 @@
+    (i j : n) :
+    toMatrixOrthonormal b f i j = ⟪b i, f (b j)⟫_𝕜 :=
+  calc
+    _ = b.repr (f (b j)) i := f.toMatrix_apply ..
+    _ = ⟪b i, f (b j)⟫_𝕜 := b.repr_apply_apply ..
+
+lemma toMatrixOrthonormal_reindex [Fintype m] [DecidableEq m]
+    (b : OrthonormalBasis m 𝕜 E) (e : m ≃ n) (f : Module.End 𝕜 E) :
+    toMatrixOrthonormal (b.reindex e) f = Matrix.reindex e e (toMatrixOrthonormal b f) :=
+  Matrix.ext fun i j =>
+    calc toMatrixOrthonormal (b.reindex e) f i j
+      _ = (b.reindex e).repr (f (b.reindex e j)) i := f.toMatrix_apply ..
+      _ = b.repr (f (b (e.symm j))) (e.symm i) := by simp
+      _ = toMatrixOrthonormal b f (e.symm i) (e.symm j) := Eq.symm (f.toMatrix_apply ..)
+
+end
+
+/-- **Don't use this definition directly.** Instead, use `Matrix.schurTriangulationBasis`,
+`Matrix.schurTriangulationUnitary`, and `Matrix.schurTriangulation`. See also
+`LinearMap.SchurTriangulationAux.of` and `Matrix.schurTriangulationAux`. -/
+structure SchurTriangulationAux (f : Module.End 𝕜 E) where
+  /-- The dimension of the inner product space `E`. -/
+  dim : ℕ
+  hdim : Module.finrank 𝕜 E = dim
+  /-- An orthonormal basis of `E` that induces an upper triangular form for `f`. -/
+  basis : OrthonormalBasis (Fin dim) 𝕜 E
+  upperTriangular : (toMatrix basis.toBasis basis.toBasis f).BlockTriangular _root_.id
+
+end
+
+/-! ## Schur's recursive triangulation procedure
+
+Given a linear endomorphism `f` on a non-trivial finite-dimensional vector space `E` over an
+algebraically closed field `𝕜`, one can always pick an eigenvalue `μ` of `f` whose corresponding
+eigenspace `V` is non-trivial. Given that `E` is also an inner product space, let `bV` and `bW` be
+othonormal bases for `V` and `Vᗮ` respectively. Then, the collection of vectors in `bV` and `bW`
+forms an othornomal basis `bE` for `E`, as the direct sum of `V` and `Vᗮ` is an internal
+decomposition of `E`. The matrix representation of `f` with respect to `bE` satisfies
+$$
+\sideset{_\mathrm{bE}}{_\mathrm{bE}}{[f]} =
+\begin{bmatrix}
+  \sideset{_\mathrm{bV}}{_\mathrm{bV}}{[f]} &
+  \sideset{_\mathrm{bW}}{_\mathrm{bV}}{[f]} \\
+  \sideset{_\mathrm{bV}}{_\mathrm{bW}}{[f]} &
+  \sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}
+\end{bmatrix} =
+\begin{bmatrix} \mu I & □ \\ 0 & \sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]} \end{bmatrix},
+$$
+which is upper triangular as long as $\sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}$ is. Finally, one
+observes that the recursion from $\sideset{_\mathrm{bE}}{_\mathrm{bE}}{[f]}$ to
+$\sideset{_\mathrm{bW}}{_\mathrm{bW}}{[f]}$ is well-founded, as the dimension of `bW` is smaller
+than that of `bE` because `bV` is non-trivial.
+
+However, in order to leverage `DirectSum.IsInternal.collectedOrthonormalBasis`, the type
+`Σ b, cond b (Fin m) (Fin n)` has to be used instead of the more natural `Fin m ⊕ Fin n` while their
+equivalence is propositionally established by `Equiv.sumEquivSigmalCond`.
```

Comment:
Highlighting this for other reviewers, since I was confused by the `Σ b, cond b m n` stuff above.